packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -28,6 +28,9 @@ const isTruthyLiteral = (type: ts.Type): boolean =>
 const isPossiblyFalsy = (type: ts.Type): boolean =>
   tsutils
     .unionTypeParts(type)
+    // Intersections like `string & {}` can also be possibly falsy,
+    // requiring us to look into the intersection.
+    .flatMap(type => tsutils.intersectionTypeParts(type))
```

Comment:
This works because:
- if it is not an intersection, `intersectionTypeParts(type)` just returns the original type, making the function behave like before
- if it is something like `string & {}`, `intersectionTypeParts(type)` returns an array containing the types `string` and `{}`. `string` in this example is possibly falsy, making our function correctly report that the entire thing is possibly falsy
- if we have something like `string & number`, the type checker directly reports this type as `never`, i.e. here we shouldn't even see that this was ever defined as an intersection in the first place